### PR TITLE
Overhauled file-parsing engine.

### PR DESCRIPTION
**Pull request by jeffwright13**
Code to address [issue 7](https://github.com/jeffwright13/audio_program_generator/issues/7)
Text file processing more robust now
Code is simplified

### CODE REVIEW

File: audio_program_generator/apg.py [hunk 1]
```diff
@@ -99,43 +108,30 @@ def gen_speech(self):
...
-                speech = AudioSegment.from_file(file, format="mp3")
-                combined += speech
-                silence = AudioSegment.silent(duration=1000 * int(interval))
-                combined += silence
+        combined = AudioSegment.empty()
+
+        for phrase, duration in tqdm(parse_textfile(self.phrase_file)):
+
+            # gTTS throws exception if given nothin-string
+            if phrase.strip() == "":
```
Comment (bbelderbos) [inline audio_program_generator/apg.py]: `if not phrase.strip():`

File: audio_program_generator/apg.py [hunk 2]
```diff
@@ -68,14 +69,22 @@
...
-    with open(filename, "r"):
-        numlines += 1
-    return numlines
+def parse_textfile(filename: str="") -> list:
+    def clean(input: str="") -> str:
+        cleaner = r"[^A-Za-z0-9\s;\v]"
+        clean = re.compile(cleaner, flags=re.MULTILINE | re.UNICODE)
+        return re.sub(clean, '', input)
+
+    def capture(cleaned: str="") -> list:
```
Comment (bbelderbos) [inline audio_program_generator/apg.py]: Capture what?

File: audio_program_generator/apg.py [hunk 3]
```diff
@@ -68,14 +69,22 @@
...
-def num_lines_in_file(filename: Path) -> int:
-    """
-    Takes text filename ; returns the number of lines in a file
-    """
-    numlines = 1
-    with open(filename, "r"):
-        numlines += 1
-    return numlines
+def parse_textfile(filename: str="") -> list:
+    def clean(input: str="") -> str:
```
Comment (bbelderbos) [inline audio_program_generator/apg.py]: Any reason to nest functions? Decorators crossed my mind ...

File: audio_program_generator/apg.py [hunk 4]
```diff
@@ -99,43 +108,30 @@ def gen_speech(self):
...
+        combined = AudioSegment.empty()
+
+        for phrase, duration in tqdm(parse_textfile(self.phrase_file)):
+
+            # gTTS throws exception if given nothin-string
+            if phrase.strip() == "":
+                continue
+
+            # Cache genearted gTTTS snippets and reuse if already present
+            Path.mkdir(Path.cwd() / ".cache") if not Path(
```
Comment (bbelderbos) [inline audio_program_generator/apg.py:121]: That `None` does not get assigned right?

I think this is more readable:

```
cache_dir = Path.cwd() / ".cache"
if not cache_dir.exists():
    cache_dir.mkdir()
```

File: audio_program_generator/apg.py [hunk 5]
```diff
@@ -99,43 +108,30 @@ def gen_speech(self):
...
+                Path.cwd() / ".cache"
+            ).exists() else None
+
+            file = Path.cwd() / ".cache" / (phrase + ".mp3")
```
Comment (bbelderbos) [inline audio_program_generator/apg.py:125]: f"{phrase}.mp3", and with above you can reuse `cache_dir` variable here

File: audio_program_generator/apg.py [hunk 6]
```diff
@@ -99,43 +108,30 @@ def gen_speech(self):
...
-
-                # Add the current speech snippet + corresponding silence
-                # to the combined file, building up for each new line.
-                speech = AudioSegment.from_file(file, format="mp3")
-                combined += speech
-                silence = AudioSegment.silent(duration=1000 * int(interval))
-                combined += silence
+        combined = AudioSegment.empty()
+
+        for phrase, duration in tqdm(parse_textfile(self.phrase_file)):
```
Comment (bbelderbos) [inline audio_program_generator/apg.py:113]: `tqdm` is cool!

**Review by bbelderbos — APPROVED** [review]

File: audio_program_generator/apg.py [hunk 7]
```diff
@@ -178,7 +178,7 @@ def main():
     if to_mix and not sound_file:
         sys.exit("Sound file " + sound_file + " does not exist. Quitting.")
 
-    A = Apg(
+    A = AudioProgramGenerator(
```
Comment (bbelderbos) [inline audio_program_generator/apg.py:182]: yep better and now `A` can be called `apg`

**Review by bbelderbos — APPROVED** [review]

File: audio_program_generator/apg.py (hunk 3 above)
Comment (jeffwright13) [reply audio_program_generator/apg.py]: As discussed in meeting, this is a 2-pass regex-based function to clean up user input and then extract ("capture") text phrases and durations. 

First is a negated character class, which excludes all the characters we don't want:
`[^A-Za-z0-9\s;\v]`
and then we sub() all of those with nothing `''`.

Next is a regex that anchors at beginning and end of line (to match whole lines), spaces are outside the capture groups, and the first capture group is non-greedy to not capture the spaces on either side:
`^\s*([\w\s]+?)\s*;\s*(\d+)\s*$`

File: audio_program_generator/apg.py (hunk 2 above)
Comment (jeffwright13) [reply audio_program_generator/apg.py]: Ok,, "extract" might be a better name, bit as it is an internal function perhaps it does not matter so much :-)

File: audio_program_generator/apg.py (hunk 4 above)
Comment (jeffwright13) [reply audio_program_generator/apg.py:121]: May entirely remove cache feature, or at least [make it configurable](https://github.com/jeffwright13/audio_program_generator/issues/11) (default off), so will address then.

File: audio_program_generator/apg.py (hunk 5 above)
Comment (jeffwright13) [reply audio_program_generator/apg.py:125]: Will implement when addressing [cache enhancement](https://github.com/jeffwright13/audio_program_generator/issues/11).

File: audio_program_generator/apg.py (hunk 3 above)
Comment (bbelderbos) [reply audio_program_generator/apg.py]: Cool, love regexes, very powerful!

**Pull request merged**